File: sdk/node/src/http.ts
/**
 * HTTP client for Brikk SDK with retry logic and error handling.
 */

import { createHmac } from 'crypto';
import {
  AuthError,
  HTTPError,
  NotFoundError,
  RateLimitError,
  ServerError,
  TimeoutError,
  ValidationError,
} from './errors.js';

export interface HTTPClientConfig {
  baseUrl: string;
  apiKey?: string;
  signingSecret?: string;
  timeoutMs?: number;
  maxRetries?: number;
}

export class HTTPClient {
  private baseUrl: string;
  private apiKey?: string;
  private signingSecret?: string;
  private timeoutMs: number;
  private maxRetries: number;

  constructor(config: HTTPClientConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, '');
    this.apiKey = config.apiKey;
    this.signingSecret = config.signingSecret;
    this.timeoutMs = config.timeoutMs || 30000;
    this.maxRetries = config.maxRetries || 3;
  }

  private buildUrl(path: string): string {
    if (path.startsWith('http')) {
      return path;
    }
    return `${this.baseUrl}/${path.replace(/^\//, '')}`;
  }

  private generateHmacSignature(timestamp: string, body: string): string {
    if (!this.signingSecret) {
      throw new AuthError('Signing secret required for HMAC authentication');
    }
    const message = `${timestamp}.${body}`;
    return createHmac('sha256', this.signingSecret)
      .update(message)
      .digest('hex');
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    let data: any;
    try {
      data = await response.json();
    } catch {
      data = { text: await response.text() };
    }

    if (response.ok) {
      return data as T;
    }

    const errorMsg = data.error || `HTTP ${response.status}`;

    if (response.status === 400) {
      throw new ValidationError(errorMsg, response.status, response);
    } else if (response.status === 401 || response.status === 403) {
      throw new AuthError(errorMsg, response.status, response);
    } else if (response.status === 404) {
      throw new NotFoundError(errorMsg, response.status, response);
    } else if (response.status === 429) {
      throw new RateLimitError(errorMsg, response.status, response);
    } else if (response.status >= 500) {
      throw new ServerError(errorMsg, response.status, response);
    } else {
      throw new HTTPError(errorMsg, response.status, response);
    }
  }

  private async sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  async request<T>(
    method: string,
    path: string,
    options: {
      body?: any;
      params?: Record<string, any>;
      headers?: Record<string, string>;
      useHmac?: boolean;
    } = {}
  ): Promise<T> {
    const url = new URL(this.buildUrl(path));
    
    if (options.params) {
      Object.entries(options.params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    const headers: Record<string, string> = options.headers || {};

    // Add authentication headers
    if (options.useHmac && this.signingSecret) {
      const timestamp = Math.floor(Date.now() / 1000).toString();
      const body = options.body ? JSON.stringify(options.body) : '';
      const signature = this.generateHmacSignature(timestamp, body);
      headers['X-Brikk-Key'] = this.apiKey || '';
      headers['X-Brikk-Timestamp'] = timestamp;
      headers['X-Brikk-Signature'] = signature;
    } else if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    if (options.body) {
      headers['Content-Type'] = 'application/json';
    }

    let lastError: Error | null = null;
    
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);

        const response = await fetch(url.toString(), {
          method,
          headers,
          body: options.body ? JSON.stringify(options.body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);
        return await this.handleResponse<T>(response);
      } catch (error: any) {
        lastError = error;

        // Don't retry on client errors (4xx except 429)
        if (error.statusCode && error.statusCode >= 400 && error.statusCode < 500 && error.statusCode !== 429) {
          throw error;
        }

        // Don't retry on last attempt
        if (attempt === this.maxRetries) {
          break;
        }

        // Exponential backoff
        const backoffMs = Math.min(1000 * Math.pow(2, attempt), 10000);
        await this.sleep(backoffMs);
      }
    }

    if (lastError?.name === 'AbortError') {
      throw new TimeoutError(`Request timed out after ${this.timeoutMs}ms`);
    }

    throw lastError || new HTTPError('Request failed');
  }

  async get<T>(path: string, params?: Record<string, any>): Promise<T> {
    return this.request<T>('GET', path, { params });
  }

  async post<T>(path: string, body?: any, useHmac = false): Promise<T> {
    return this.request<T>('POST', path, { body, useHmac });
  }

  async put<T>(path: string, body?: any): Promise<T> {
    return this.request<T>('PUT', path, { body });
  }

  async delete<T>(path: string): Promise<T> {
    return this.request<T>('DELETE', path);
  }
}

File: sdk/python/brikk/_http.py
# -*- coding: utf-8 -*-
"""Synchronous HTTP client for Brikk SDK."""

import time
import hmac
import hashlib
import json
from typing import Any, Dict, Optional
from urllib.parse import urljoin

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

from .errors import (
    AuthError,
    HTTPError,
    NotFoundError,
    RateLimitError,
    ServerError,
    TimeoutError,
    ValidationError,
)


class HTTPClient:
    """Synchronous HTTP client with retry logic and error handling."""

    def __init__(
        self,
        base_url: str,
        api_key: Optional[str] = None,
        signing_secret: Optional[str] = None,
        timeout: int = 30,
        max_retries: int = 3,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.signing_secret = signing_secret
        self.timeout = timeout

        # Create session with retry logic
        self.session = requests.Session()
        retry_strategy = Retry(
            total=max_retries,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["GET", "POST", "PUT", "PATCH", "DELETE"],
            backoff_factor=1,
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)

    def _build_url(self, path: str) -> str:
        """Build full URL from path."""
        if path.startswith("http"):
            return path
        return urljoin(self.base_url + "/", path.lstrip("/"))

    def _generate_hmac_signature(self, timestamp: str, body: str) -> str:
        """Generate HMAC-SHA256 signature for request."""
        if not self.signing_secret:
            raise AuthError("Signing secret required for HMAC authentication")
        message = f"{timestamp}.{body}".encode("utf-8")
        return hmac.new(
            self.signing_secret.encode("utf-8"),
            message,
            hashlib.sha256,
        ).hexdigest()

    def _handle_response(self, response: requests.Response) -> Dict[str, Any]:
        """Handle HTTP response and raise appropriate errors."""
        try:
            data = response.json() if response.content else {}
        except json.JSONDecodeError:
            data = {"text": response.text}

        if response.status_code == 200 or response.status_code == 201:
            return data

        error_msg = data.get("error", f"HTTP {response.status_code}")

        if response.status_code == 400:
            raise ValidationError(error_msg, response.status_code, response)
        elif response.status_code == 401 or response.status_code == 403:
            raise AuthError(error_msg, response.status_code, response)
        elif response.status_code == 404:
            raise NotFoundError(error_msg, response.status_code, response)
        elif response.status_code == 429:
            raise RateLimitError(error_msg, response.status_code, response)
        elif response.status_code >= 500:
            raise ServerError(error_msg, response.status_code, response)
        else:
            raise HTTPError(error_msg, response.status_code, response)

    def request(
        self,
        method: str,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        use_hmac: bool = False,
    ) -> Dict[str, Any]:
        """Make HTTP request with error handling and retries."""
        url = self._build_url(path)
        request_headers = headers or {}

        # Add authentication headers
        if use_hmac and self.signing_secret:
            timestamp = str(int(time.time()))
            body = json.dumps(json_data) if json_data else ""
            signature = self._generate_hmac_signature(timestamp, body)
            request_headers.update({
                "X-Brikk-Key": self.api_key or "",
                "X-Brikk-Timestamp": timestamp,
                "X-Brikk-Signature": signature,
            })
        elif self.api_key:
            request_headers["Authorization"] = f"Bearer {self.api_key}"

        if json_data:
            request_headers["Content-Type"] = "application/json"

        try:
            response = self.session.request(
                method=method,
                url=url,
                json=json_data,
                params=params,
                headers=request_headers,
                timeout=self.timeout,
            )
            return self._handle_response(response)
        except requests.Timeout as e:
            raise TimeoutError(f"Request timed out after {self.timeout}s") from e
        except requests.RequestException as e:
            raise HTTPError(f"Request failed: {str(e)}") from e

    def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make GET request."""
        return self.request("GET", path, params=params)

    def post(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        use_hmac: bool = False,
    ) -> Dict[str, Any]:
        """Make POST request."""
        return self.request("POST", path, json_data=json_data, use_hmac=use_hmac)

    def put(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Make PUT request."""
        return self.request("PUT", path, json_data=json_data)

    def delete(self, path: str) -> Dict[str, Any]:
        """Make DELETE request."""
        return self.request("DELETE", path)

